File: cpp/cpp-boost/exer06b01-blocking-queue.cpp
/*
BLOCKING QUEUE IMPLEMENTATION
Version B01: General blocking queues
             Underlying mechanism: Semaphores
*/


#include <iostream>
#include <queue>
#include <string>
#include <stdexcept>
#include <boost/chrono.hpp>
#include <boost/thread.hpp>
#include "mylib-semaphore.hpp"
using namespace std;



template <typename T>
class BlockingQueue {

private:
    int capacity;

    mylib::Semaphore semRemain;
    mylib::Semaphore semFill;
    boost::mutex mut;

    std::queue<T> q;


public:
    BlockingQueue(int capacity) : capacity(capacity), semRemain(capacity), semFill(0) {
        if (this->capacity <= 0)
            throw std::invalid_argument("capacity must be a positive integer");
    }


    void put(const T& value) {
        semRemain.acquire();

        {
            boost::unique_lock<boost::mutex> lk(mut);
            q.push(value);
        }

        semFill.release();
    }


    T take() {
        T result;
        semFill.acquire();

        {
            boost::unique_lock<boost::mutex> lk(mut);
            result = q.front();
            q.pop();
        }

        semRemain.release();
        return result;
    }

};



void producer(BlockingQueue<std::string>* blkQueue) {
    std::string arr[] = { "nice", "to", "meet", "you" };

    for (int i = 0; i < 4; ++i) {
        std::string& data = arr[i];
        cout << "Producer: " << data << endl;
        blkQueue->put(data);
        cout << "Producer: " << data << "\t\t\t[done]" << endl;
    }
}



void consumer(BlockingQueue<std::string>* blkQueue) {
    std::string data;
    boost::this_thread::sleep_for(boost::chrono::seconds(5));

    for (int i = 0; i < 4; ++i) {
        data = blkQueue->take();
        cout << "\tConsumer: " << data << endl;

        if (0 == i)
            boost::this_thread::sleep_for(boost::chrono::seconds(5));
    }
}



int main() {
    BlockingQueue<std::string> blkQueue(2); // capacity = 2

    boost::thread thProducer(&producer, &blkQueue);
    boost::thread thConsumer(&consumer, &blkQueue);

    thProducer.join();
    thConsumer.join();

    return 0;
}

File: python/exer06b01_blocking_queue.py
'''
BLOCKING QUEUE IMPLEMENTATION
Version B01: General blocking queues
             Underlying mechanism: Semaphores
'''

import time
import threading



class BlockingQueue:
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError('capacity must be a positive integer')
        # self.__capacity = capacity
        self.__sem_remain = threading.Semaphore(capacity)
        self.__sem_fill = threading.Semaphore(0)
        self.__lk = threading.Lock()
        self.__q = [] # queue


    def put(self, value):
        self.__sem_remain.acquire()

        with self.__lk:
            self.__q.append(value)

        self.__sem_fill.release()


    def take(self):
        self.__sem_fill.acquire()

        with self.__lk:
            result = self.__q.pop(0)

        self.__sem_remain.release()
        return result



def producer(q: BlockingQueue):
    arr = [ 'nice', 'to', 'meet', 'you' ]

    for data in arr:
        print(f'Producer: {data}')
        q.put(data)
        print(f'Producer: {data} \t\t\t[done]')



def consumer(q: BlockingQueue):
    time.sleep(5)

    for i in range(4):
        data = q.take()
        print(f'\tConsumer: {data}')

        if i == 0:
            time.sleep(5)



blkqueue = BlockingQueue(2) # capacity = 2
threading.Thread(target=producer, args=(blkqueue,)).start()
threading.Thread(target=consumer, args=(blkqueue,)).start()
